### DISCUSSION TITLE
Reaction not firing - what am I missing?

### DISCUSSION THREAD

**Discussion by scottwillman** [mobxjs/mobx#2553] [Q&A] [answered]
This reaction isn't firing and I don't have a deep enough understanding yet as to why. Can someone spot my mistake?

```
import { autorun, reaction, makeAutoObservable } from "mobx";

class MyStore {
  cache = new Map();

  constructor() {
    makeAutoObservable(this);

    // this logs
    autorun(() => {
      console.log(this.cache.toJSON());
    });

    // this does not
    reaction(
      () => this.cache,
      (value, prevValue) => {
        console.log(prevValue);
        console.log(value);
      }
    );
  }

  addToCache(key, value) {
    this.cache.set(key, value);
  }
}

const store = new MyStore();
store.addToCache("foo", "bar");
store.addToCache("sing", "song");
```
https://codesandbox.io/s/cranky-mahavira-63ttc?file=/src/App.js

**Comment by scottwillman** [comment]
It looks like I needed to access something specific about the Map, not just the root map itself. This works:

```
    reaction(
      () => this.cache.entries(),
      (value, prevValue) => {
        console.log(prevValue);
        console.log(value);
      }
    );
```

**Reply by danielkcz** [reply]
That is correct. If you need to watch over the whole thing, you can use `toJS`, but that generally beats the purpose of reactivity to "react" to changes of things you need.

**Comment by urugator** [comment, marked as answer]
[This could help](https://github.com/mobxjs/mobx/issues/2222#issuecomment-561233503) (you can skip first 4 paragraphs)

**Reply by scottwillman** [reply]
That's a really great explanation, thanks for the link. Ironically, creating a cache and trying to find the diffs was exactly how I was thinking at the time. So, thanks for this recalibration.
